jujugui/static/gui/src/app/store/env/controller-api.js

```diff
@@ -0,0 +1,683 @@
+     * @return {undefined} Dispatches only.
+     */
+    dispatch_result: function(data) {
+      var tid = data['request-id'];
+      if (tid in this._txn_callbacks) {
+        this._txn_callbacks[tid].call(this, data);
+        delete this._txn_callbacks[tid];
+      }
+    },
+
+    /**
+     * Send a message to the server using the WebSocket connection.
+     *
+     * @method _send_rpc
+     * @private
+     * @param {Object} op The operation to perform (compatible with the
+         juju-core format specification, see "/doc/draft/api.txt" in
+         lp:~rogpeppe/juju-core/212-api-doc).
+     * @param {Function} callback A callable that must be called once the
+         backend returns results.
+     * @return {undefined} Sends a message to the server only.
+     */
+    _send_rpc: function(op, callback) {
+      var facade = op.type;
+      // The facades info is only available after logging in (as the facades
+      // are sent as part of the login response). For this reason, do not
+      // check if the "Admin" facade is supported, but just assume it is,
+      // otherwise even logging in ("Admin.Login") would be impossible.
+      var version = op.version;
+      if (facade !== 'Admin') {
+        version = this.findFacadeVersion(facade, version);
+      }
+      if (version === null) {
+        var err = 'api client: operation not supported: ' + JSON.stringify(op);
+        console.error(err);
+        if (callback) {
+          callback({error: err});
+        }
+        return;
+      }
+      if (this.ws.readyState !== 1) {
+        console.log(
+          'Websocket is not open, dropping request. ' +
+          'readyState: ' + this.ws.readyState, op);
+        return;
+      }
+      op.version = version;
+      var tid = this._counter += 1;
+      if (callback) {
+        this._txn_callbacks[tid] = callback;
+      }
+      op['request-id'] = tid;
+      if (!op.params) {
+        op.params = {};
+      }
+      var msg = JSON.stringify(op);
+      this.ws.send(msg);
+    },
+
+    /**
+     * React to the results of sending a login message to the server.
+     *
+     * @method handleLogin
+     * @param {Object} data The response returned by the server.
+     * @return {undefined} Nothing.
+     */
+    handleLogin: function(data) {
+      this.pendingLoginResponse = false;
+      this.userIsAuthenticated = !data.error;
+      if (this.userIsAuthenticated) {
+        var response = data.response;
+        // If login succeeded store the facades and user information, and
+        // retrieve model info.
+        var facadeList = response.facades || [];
+        var facades = facadeList.reduce(function(previous, current) {
+          previous[current.name] = current.versions;
+          return previous;
+        }, {});
+        this.set('facades', facades);
+        var userInfo = response['user-info'];
+        this.set('readOnly', !!userInfo['read-only']);
+        this.set('serverTag', response['server-tag']);
```

Comment:
This login method doesn't start the watcher or attempt to fetch the model info as that's not supported on the controller endpoint